// @gala-chain/api/lib/src/client/api/CommonContractAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commonContractAPI = void 0;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const types_1 = require("../../types");
const commonContractAPI = (client) => ({
    async GetContractVersion() {
        const resp = await client.evaluateTransaction("GetContractVersion");
        return resp;
    },
    async GetContractAPI() {
        const resp = await client.evaluateTransaction("GetContractAPI");
        return resp;
    },
    async GetObjectByKey(key) {
        const dto = await (0, types_1.createValidDTO)(types_1.GetObjectDto, { objectId: key });
        const resp = await client.evaluateTransaction("GetObjectByKey", dto);
        return resp;
    },
    async GetObjectHistory(key) {
        const dto = await (0, types_1.createValidDTO)(types_1.GetObjectHistoryDto, { objectId: key });
        const resp = await client.evaluateTransaction("GetObjectHistory", dto);
        return resp;
    },
    async DryRun(method, callerPublicKey, dto) {
        const dryRunDto = await (0, types_1.createValidDTO)(types_1.DryRunDto, { method, callerPublicKey, dto });
        const resp = await client.evaluateTransaction("DryRun", dryRunDto);
        return resp;
    },
    async BatchSubmit(batch) {
        const resp = await client.submitTransaction("BatchSubmit", batch);
        return resp;
    },
    async BatchEvaluate(batch) {
        const resp = await client.evaluateTransaction("BatchEvaluate", batch);
        return resp;
    }
});
exports.commonContractAPI = commonContractAPI;
