// @gala-chain/api/lib/src/client/api/PublicKeyContractAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.publicKeyContractAPI = void 0;
/*
 * Copyright (c) Gala Games Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const types_1 = require("../../types");
const CommonContractAPI_1 = require("./CommonContractAPI");
const publicKeyContractAPI = (client) => ({
    ...(0, CommonContractAPI_1.commonContractAPI)(client),
    GetPublicKey(user) {
        if (typeof user === "string") {
            const dto = new types_1.GetPublicKeyDto();
            dto.user = (0, types_1.asValidUserRef)(user);
            return client.evaluateTransaction("GetPublicKey", dto, types_1.PublicKey);
        }
        else {
            return client.evaluateTransaction("GetPublicKey", user !== null && user !== void 0 ? user : new types_1.GetPublicKeyDto(), types_1.PublicKey);
        }
    },
    RegisterUser(dto) {
        return client.submitTransaction("RegisterUser", dto);
    },
    RegisterEthUser(dto) {
        return client.submitTransaction("RegisterEthUser", dto);
    },
    UpdatePublicKey(dto) {
        return client.submitTransaction("UpdatePublicKey", dto);
    },
    GetMyProfile(dto) {
        return client.evaluateTransaction("GetMyProfile", dto, types_1.UserProfile);
    }
});
exports.publicKeyContractAPI = publicKeyContractAPI;
