// @gala-chain/connect/lib/src/types/galaChain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GalaChainResponseError = exports.GalaChainResponseSuccess = void 0;
/**
 * Represents a successful response from a GalaChain operation.
 * @template T - The type of data returned in the response
 */
class GalaChainResponseSuccess {
    /**
     * Creates a new successful response.
     * @param data - The success response data
     * @param hash - Optional transaction hash
     */
    constructor(data, hash) {
        this.Hash = hash;
        this.Status = data.Status;
        this.Data = data.Data;
    }
}
exports.GalaChainResponseSuccess = GalaChainResponseSuccess;
/**
 * Represents an error response from a GalaChain operation.
 * @template T - The type of data associated with the error
 */
class GalaChainResponseError {
    /**
     * Creates a new error response.
     * @param data - The error data to construct the response from
     */
    constructor(data) {
        var _a;
        if (typeof data.error === "string" || !data.error) {
            this.Error = data.error;
            this.Message = data.message;
            this.ErrorCode = (_a = data.statusCode) !== null && _a !== void 0 ? _a : 500;
        }
        else {
            this.Error = data.error.ErrorKey;
            this.Message = data.message;
            this.ErrorCode = data.error.ErrorCode;
        }
    }
}
exports.GalaChainResponseError = GalaChainResponseError;
