// class-validator-jsonschema/build/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.targetConstructorToSchema = exports.validationMetadataArrayToSchemas = exports.validationMetadatasToSchemas = exports.JSONSchema = void 0;
const tslib_1 = require("tslib");
const cv = tslib_1.__importStar(require("class-validator"));
const lodash_groupby_1 = tslib_1.__importDefault(require("lodash.groupby"));
const lodash_merge_1 = tslib_1.__importDefault(require("lodash.merge"));
const decorators_1 = require("./decorators");
const defaultConverters_1 = require("./defaultConverters");
const options_1 = require("./options");
var decorators_2 = require("./decorators");
Object.defineProperty(exports, "JSONSchema", { enumerable: true, get: function () { return decorators_2.JSONSchema; } });
function validationMetadatasToSchemas(userOptions) {
    const options = Object.assign(Object.assign({}, options_1.defaultOptions), userOptions);
    const metadatas = getMetadatasFromStorage(options.classValidatorMetadataStorage);
    return validationMetadataArrayToSchemas(metadatas, userOptions);
}
exports.validationMetadatasToSchemas = validationMetadatasToSchemas;
function validationMetadataArrayToSchemas(metadatas, userOptions) {
    const options = Object.assign(Object.assign({}, options_1.defaultOptions), userOptions);
    const schemas = {};
    Object.entries((0, lodash_groupby_1.default)(metadatas, ({ target }) => {
        var _a;
        return (_a = target[options.schemaNameField]) !== null && _a !== void 0 ? _a : target.name;
    })).forEach(([key, ownMetas]) => {
        const target = ownMetas[0].target;
        const metas = ownMetas
            .concat(getInheritedMetadatas(target, metadatas))
            .filter((propMeta) => !(isExcluded(propMeta, options) ||
            isExcluded(Object.assign(Object.assign({}, propMeta), { target }), options)));
        const properties = {};
        Object.entries((0, lodash_groupby_1.default)(metas, 'propertyName')).forEach(([propName, propMetas]) => {
            const schema = applyConverters(propMetas, options);
            properties[propName] = applyDecorators(schema, target, options, propName);
        });
        const definitionSchema = {
            properties,
            type: 'object',
        };
        const required = getRequiredPropNames(target, metas, options);
        if (required.length > 0) {
            definitionSchema.required = required;
        }
        schemas[key] = applyDecorators(definitionSchema, target, options, target.name);
    });
    return schemas;
}
exports.validationMetadataArrayToSchemas = validationMetadataArrayToSchemas;
function getTargetConstructorSchema(schemas, targetConstructor) {
    if (!targetConstructor.name) {
        return {};
    }
    else if (schemas[targetConstructor.name]) {
        return schemas[targetConstructor.name];
    }
    else {
        return getTargetConstructorSchema(schemas, Object.getPrototypeOf(targetConstructor));
    }
}
function targetConstructorToSchema(targetConstructor, userOptions) {
    const options = Object.assign(Object.assign({}, options_1.defaultOptions), userOptions);
    const storage = options.classValidatorMetadataStorage;
    let metadatas = storage.getTargetValidationMetadatas(targetConstructor, '', true, false);
    metadatas = populateMetadatasWithConstraints(storage, metadatas);
    const schemas = validationMetadataArrayToSchemas(metadatas, userOptions);
    return getTargetConstructorSchema(schemas, targetConstructor);
}
exports.targetConstructorToSchema = targetConstructorToSchema;
function getMetadatasFromStorage(storage) {
    const metadatas = [];
    for (const value of storage.validationMetadatas) {
        metadatas.push(...populateMetadatasWithConstraints(storage, value[1]));
    }
    return metadatas;
}
function populateMetadatasWithConstraints(storage, metadatas) {
    return metadatas.map((meta) => {
        if (meta.constraintCls) {
            const constraint = storage.getTargetValidatorConstraints(meta.constraintCls);
            if (constraint.length > 0) {
                return Object.assign(Object.assign({}, meta), { type: constraint[0].name });
            }
        }
        return Object.assign({}, meta);
    });
}
function getInheritedMetadatas(target, metadatas) {
    return metadatas.filter((d) => d.target instanceof Function &&
        target.prototype instanceof d.target &&
        !metadatas.find((m) => m.propertyName === d.propertyName &&
            m.target === target &&
            m.type === d.type));
}
function applyConverters(propertyMetadatas, options) {
    const converters = Object.assign(Object.assign({}, defaultConverters_1.defaultConverters), options.additionalConverters);
    const convert = (meta) => {
        var _a;
        const typeMeta = (_a = options.classTransformerMetadataStorage) === null || _a === void 0 ? void 0 : _a.findTypeMetadata(meta.target, meta.propertyName);
        const isMap = typeMeta &&
            typeMeta.reflectedType &&
            new typeMeta.reflectedType() instanceof Map;
        const converter = converters[meta.type] || converters[cv.ValidationTypes.CUSTOM_VALIDATION];
        const items = typeof converter === 'function' ? converter(meta, options) : converter;
        if (meta.each && isMap) {
            return {
                additionalProperties: Object.assign({}, items),
                type: 'object',
            };
        }
        return meta.each ? { items, type: 'array' } : items;
    };
    return (0, lodash_merge_1.default)({}, ...propertyMetadatas.map(convert));
}
function isExcluded(propertyMetadata, options) {
    var _a;
    return !!((_a = options.classTransformerMetadataStorage) === null || _a === void 0 ? void 0 : _a.findExcludeMetadata(propertyMetadata.target, propertyMetadata.propertyName));
}
function applyDecorators(schema, target, options, propertyName) {
    const additionalSchema = (0, decorators_1.getMetadataSchema)(target.prototype, propertyName);
    return typeof additionalSchema === 'function'
        ? additionalSchema(schema, options)
        : (0, lodash_merge_1.default)({}, schema, additionalSchema);
}
function getRequiredPropNames(target, metadatas, options) {
    function isDefined(metas) {
        return (metas && metas.some(({ type }) => type === cv.ValidationTypes.IS_DEFINED));
    }
    function isOptional(metas) {
        return (metas &&
            metas.some(({ type }) => [cv.ValidationTypes.CONDITIONAL_VALIDATION, cv.IS_EMPTY].includes(type)));
    }
    return Object.entries((0, lodash_groupby_1.default)(metadatas, (m) => m.propertyName))
        .filter(([_, metas]) => {
        const own = metas.filter((m) => m.target === target);
        const inherited = metas.filter((m) => m.target !== target);
        return options.skipMissingProperties
            ? isDefined(own) || (!isOptional(own) && isDefined(inherited))
            : !(isOptional(own) || isOptional(inherited));
    })
        .map(([name]) => name);
}
